driver.go

```diff
@@ -47,10 +47,18 @@ import (
 // Driver is exported to allow it to be used directly.
 type Driver struct{}
 
-// Open a Connection to the server.
-// See https://github.com/apache/calcite-avatica-go#dsn for more information
-// on how the DSN is formatted.
-func (a *Driver) Open(dsn string) (driver.Conn, error) {
+// Connector implements the driver.Connector interface
+type Connector struct {
+	additionalInfos   map[string]string
+	additionalHeaders map[string]string
```

Comment:
Instead of `additionalHeaders`, allow the user of the library to supply in their own `http.Client`, this property should be public.

Reply:
I see this being even more usefull, but then the existing auth code would basically needs to be removed since it would touch the supplied http client (setting the transport).

I could leave it as it is, and only overwrite the transport if no http client was passed by the user. What do you think, would a user be confused if he supplies a http client and specifies for example the kerberos settings and things wouldn't work?

IMO with that option to pass in a http client it would just be consequently to not do any auth stuff within the driver.

Reply:
That sounds like a good idea, if no http client is supplied, we can instantiate a new client (from the existing code). If it is supplied, we do not modify the http client at all.

To improve the user experience, we can extract these 2 branches into public functions that takes a `http.Client` and makes changes to them: `AddDigestAuthenticationCredentialsToHTTPClient` and `AddKerberosCredentialsToHTTPClient`. That way, anyone who supplies his own http client gets to use digest or kerberos auth without having to write most of the code themselves.

Reply:
I refactored out the different authentication methods, so now you can setup the http.Client as you want, and pass it into the Connector.